base/runtime_internals.jl

```diff
@@ -173,6 +173,53 @@ false
 """
 ispublic(m::Module, s::Symbol) = ccall(:jl_module_public_p, Cint, (Any, Any), m, s) != 0
 
+"""
+    @__FUNCTION__ -> Function
+
+Get the innermost enclosing function object.
+
+!!! note
+    In functions like `f() = [@__FUNCTION__ for _ in 1:10]`, this would
+    refer to the generator function in the comprehension, NOT the enclosing
+    function `f`. Similarly, in a closure function, this will refer to the
+    closure function object rather than the enclosing function.
+
+!!! note
+    This does not work in the context of callable structs as there is no
+    function to refer to, and will result in an `UndefVarError`.
+
+# Examples
+
+`@__FUNCTION__` is useful for closures that need to refer to themselves,
+as otherwise the function object itself would be a variable and would be boxed:
+
+```jldoctest
+julia> function make_fib()
+           fib(n) = n <= 1 ? 1 : (@__FUNCTION__)(n - 1) + (@__FUNCTION__)(n - 2)
+           return fib
+       end
+make_fib (generic function with 1 method)
+
+julia> make_fib()(7)
+21
+```
+
+If we had instead written `fib(n) = n <= 1 ? 1 : fib(n - 1) + fib(n - 2)`
+for the closure function, `fib` would be treated as a variable, and be boxed.
+
+Note that `@__FUNCTION__` is available for anonymous functions:
+
+```jldoctest
+julia> factorial = n -> n <= 1 ? 1 : n * (@__FUNCTION__)(n - 1);
+
+julia> factorial(5)
+120
+```
+"""
+macro __FUNCTION__()
+    return esc(:(var"#self#"))
```

Comment:
Should probably check `isdefined` and return `nothing` if not to avoid weird errors in e.g. the REPL.

Reply:
Or perhaps returning `nothing` would be helpful for knowing whether code is inside a function or not?

Reply:
e.g., maybe something like this?

```julia
macro __FUNCTION__()
    $(esc(Expr(:isdefined, :var"#self#"))) || error("@__FUNCTION__ can only be used within a function")
    $(esc(:var"#self#"))
end
```

Reply:
Going with this^ for now; feel free to unresolve if needed